internal/fs/handle/file.go

```diff
@@ -103,6 +112,51 @@ func (fh *FileHandle) Unlock() {
 	fh.mu.Unlock()
 }
 
+// ReadWithReadManager reads data at the given offset using the read manager if available,
+// falling back to inode.Read otherwise. It may be more efficient than directly calling inode.Read.
+//
+// LOCKS_REQUIRED(fh.mu)
+// LOCKS_REQUIRED(fh.inode.mu)
+// UNLOCK_FUNCTION(fh.inode.mu)
+func (fh *FileHandle) ReadWithReadManager(ctx context.Context, dst []byte, offset int64, sequentialReadSizeMb int32) ([]byte, int, error) {
+	// Attempt to initialize or validate the read manager
+	err := fh.tryEnsureReadManager(ctx, sequentialReadSizeMb)
+	if err != nil {
+		fh.inode.Unlock()
+		return nil, 0, fmt.Errorf("tryEnsureReadManager: %w", err)
+	}
+
+	// If read manager is available, try reading through it
+	if fh.readManager != nil {
+		fh.inode.Unlock()
+
+		readerResponse, readErr := fh.readManager.ReadAt(ctx, dst, offset)
+		if errors.Is(readErr, io.EOF) {
+			// Unexpected EOF — log and propagate
+			logger.Warnf("Unexpected EOF error encountered while reading, err: %v type: %T", readErr, readErr)
+			return nil, 0, io.EOF
+		}
+		if readErr != nil {
+			return nil, 0, fmt.Errorf("fh.readManager.ReadAt: %w", readErr)
+		}
+
+		// Return data buffer (may or may not be the same as dst) and number of bytes read
+		return readerResponse.DataBuf, readerResponse.Size, nil
+	}
+
+	// If read manager is not available, fall back to reading via inode
+	defer fh.inode.Unlock()
+
+	n, err := fh.inode.Read(ctx, dst, offset)
+
+	if errors.Is(err, io.EOF) {
+		err = nil
```

Comment:
We are anyway doing it in caller of fs.go, but doing it here makes sense as for file handle reads also we are doing it in this function only. 

Let me know if you find any concerns @raj-prince .